### ISSUE TITLE
passing C preprocessor variable which is a string

### ISSUE THREAD

**Issue by pwittich** [microsoft/vscode-makefile-tools#169]
Issue Type: <b>Bug</b>

In my makefile for a c project (microcontroller code) I pass in the git hash as a string via a preprocessor directive. The quotes on the preprocessor directive are not handled properly.  Please see attached screenshot.

![vscode_bug](https://user-images.githubusercontent.com/7728573/117850078-f7d62100-b252-11eb-8beb-0e4ddb2478d0.jpg)

In the dryrun.log file the command is listed as follows:
```
arm-none-eabi-gcc -mthumb -mcpu=cortex-m4 -mfpu=fpv4-sp-d16 -mfloat-abi=hard -mfp16-format=ieee -ffunction-sections -fdata-sections -MD -std=c11 -Wall -pedantic -DPART_TM4C1290NCPDT -fstack-usage -c -g -D DEBUG -O0 -DTARGET_IS_TM4C129_RA2  -I. -I.. -I../.. -DFIRMWARE_VERSION=\""v0.31.14-2-gd05c"\" -DTARGET_IS_TM4C129_RA2 -Dgcc -o gcc/bl_main.o bl_main.c
```
and the compilation completes successfully.  I expect the macro to expand to

```
#define FIRMWARE_VERSION "v0.31.14-2-gd05c"
```
not to
```
#define FIRMWARE_VERSION \""v0.31.14-2-gd05c"\"
```

Extension version: 0.2.1
VS Code version: Code 1.56.0 (cfa2e218100323074ac1948c885448fdf4de2a7f, 2021-05-04T22:06:21.189Z)
OS version: Darwin x64 20.4.0
Remote OS version: Linux x64 3.10.0-1160.6.1.el7.x86_64

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-7600K CPU @ 3.80GHz (4 x 3800)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|4, 2, 2|
|Memory (System)|40.00GB (1.84GB free)|
|Process Argv|--crash-reporter-id c670bf7f-6611-4ff8-bc75-7eaf3fbae75d|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: lnx4189.classe.cornell.edu|
|OS|Linux x64 3.10.0-1160.6.1.el7.x86_64|
|CPUs|Intel(R) Xeon(R) CPU E5-1620 0 @ 3.60GHz (8 x 3699)|
|Memory (System)|15.43GB (3.74GB free)|
|VM|0%|
</details>

**Comment by andreeis** [comment]
@pwittich, can you define the git hash with only one pair of quotes? Why 2 pairs? Before implementing complicated scenarios of removing or not removing nested quotes (other projects use them in more complicated ways for various reasons) let us know why you can't modify your makefile and what scenario would remain uncovered by having only one pair, like:  -I../.. -DFIRMWARE_VERSION="v0.31.14-2-gd05c".

**Comment by pwittich** [comment]
Alas if I do that it does not compile.

The actual snippet of the makefile looks as follows.
```
# GIT 
GIT_VERSION="$(shell git describe --abbrev=4 --dirty --always --tags)"
CFLAGS+= -DFIRMWARE_VERSION=\"$(GIT_VERSION)\"
```
`CFLAGS` gets passed directly to the compilation command. If I don't protect the second set of quotes as written and instead write it as below:
```
CFLAGS+= -DFIRMWARE_VERSION="$(GIT_VERSION)"
```
I get the following error (identical with or without the unprotected quote around $(GIT_VERSION).)

```
arm-none-eabi-gcc [... irrelevant stuff dropped]  -DFIRMWARE_VERSION=v0.31.14-2-gd05c-dirty  -o gcc/cm_mcu.o cm_mcu.c
cm_mcu.c: In function 'gitVersion':
<command-line>: error: too many decimal points in number
cm_mcu.c:207:29: note: in expansion of macro 'FIRMWARE_VERSION'
   const char * gitVersion = FIRMWARE_VERSION BUILD_TYPE;
                             ^~~~~~~~~~~~~~~~
<command-line>: error: 'v0' undeclared (first use in this function)
cm_mcu.c:207:29: note: in expansion of macro 'FIRMWARE_VERSION'
   const char * gitVersion = FIRMWARE_VERSION BUILD_TYPE;
                             ^~~~~~~~~~~~~~~~
<command-line>: note: each undeclared identifier is reported only once for each function it appears in
cm_mcu.c:207:29: note: in expansion of macro 'FIRMWARE_VERSION'
   const char * gitVersion = FIRMWARE_VERSION BUILD_TYPE;
                             ^~~~~~~~~~~~~~~~
<command-line>: error: expected ',' or ';' before numeric constant
cm_mcu.c:207:29: note: in expansion of macro 'FIRMWARE_VERSION'
   const char * gitVersion = FIRMWARE_VERSION BUILD_TYPE;
                             ^~~~~~~~~~~~~~~~
```
i.e., FIRMWARE_VERSION is treated like some weird combination of a variable and a number. 

This is presumably because of the intermediate shell that the compilation command is run under. (I'm using various unix-like systems, i.e., Linux and MacOS.)

**Comment by andreeis** [comment]
Ok, thank you for the details. We will fix this.

**Comment by andreeis** [comment]
Unfortunately, while investigating your scenario I found more similar but different scenarios of nested quotes that need to be fixed together, with maximum attention and ideally after we implement some regression testing at least for the parsers.
This is to risky to address for 0.2.2 and I'm moving it for 0.3 (which could be 3-4 months away). As soon as I have the fix in, I can provide a vsix to unblock you before we make the official release and that should be much earlier because I plan to look thoroughly at defines and quotes as soon as I'm done with the testing infrastructure. 
I hope that you using a manually edited log will suffice and won't prevent you from using the Makefile Tools 0.2.1 extension.

**Comment by andreeis** [comment]
@pwittich, we released Makefile Tools 0.3.0 today which includes a bug fix for this problem. Let us know if you encounter any other issues.

If you would like, you can also leave us a review on [Market Place](https://marketplace.visualstudio.com/items?itemName=ms-vscode.makefile-tools&ssr=false#review-details).
